### PR TITLE
Initial Java Bindings using GenomicsDB api to create List of Variant Calls

### PR DESCRIPTION
**Pull request by nalinigans**

### CODE REVIEW

File: src/main/jni/src/genomicsdb_GenomicsDBQuery.cc [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+#define INIT(VAR,CODE)                                                                                 \
+  do {                                                                                                 \
+    VAR = CODE;                                                                                        \
+    if (VAR == NULL) {                                                                                 \
+      throw GenomicsDBException("genomicsdb_GenomicsDBQuery.cc#initialize:"+std::to_string(__LINE__)); \
+    }                                                                                                  \
+  } while (false)
+
+void initialize(JNIEnv *env) {
+  std::lock_guard<std::mutex> guard(initializing_mutex_);
```
Comment (kgururaj) [inline src/main/jni/src/genomicsdb_GenomicsDBQuery.cc]: An easier way to  do this is to create a class called JNIInitializer (for example). Constructor does the initialization. Instantiate a single global object of this class. This way when the library is loaded, the initialization is done (no need to worry about race conditions etc).

File: src/main/jni/src/genomicsdb_GenomicsDBQuery.cc [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  std::vector<std::string> result;
+  result.reserve(size);
+  for (jint i=0; i<size; i++) {
+    jstring element = static_cast<jstring>(env->CallObjectMethod(arrayList, java_ArrayList_get_, i));
+    auto element_cstr = env->GetStringUTFChars(element, nullptr);
+    result.emplace_back(element_cstr);
+    env->ReleaseStringUTFChars(element, element_cstr);
+    env->DeleteLocalRef(element);
+  }
+  return std::move(result);
```
Comment (kgururaj) [inline src/main/jni/src/genomicsdb_GenomicsDBQuery.cc]: I don't think it helps here, does it? Generally, I have seen moves help in the caller function

File: src/main/jni/src/genomicsdb_GenomicsDBQuery.cc (hunk 1 above)
Comment (nalinigans) [reply src/main/jni/src/genomicsdb_GenomicsDBQuery.cc]: Yep, I was thinking of doing it in JNI_OnLoad, but did not want  to initialize when the api was not being invoked. On the other hand, maybe not having a mutex may be beneficial.

File: src/main/jni/src/genomicsdb_GenomicsDBQuery.cc (hunk 2 above)
Comment (nalinigans) [reply src/main/jni/src/genomicsdb_GenomicsDBQuery.cc]: Yes, it is not necessary, return value optimization will do the right thing. Thanks.

File: src/main/jni/src/genomicsdb_GenomicsDBQuery.cc (hunk 1 above)
Comment (nalinigans) [reply src/main/jni/src/genomicsdb_GenomicsDBQuery.cc]: @kgururaj , used Java_org_genomicsdb_reader_GenomicsDBQuery_jniInitialize() to initialize the JNI api interface in the static constructor as you suggested. Thanks.

**Pull request merged**